src/tensors/treetransformers.jl

```diff
@@ -7,68 +7,136 @@ abstract type TreeTransformer end
 
-    rows = Int[]
-    cols = Int[]
-    vals = sectorscalartype(sectortype(Vdst))[]
+    uncoupleds_dst = map(structure_dst.fusiontreelist) do (f₁, f₂)
+        return TupleTools.vcat(f₁.uncoupled, f₂.uncoupled)
+    end
+
+    T = sectorscalartype(I)
+    N = numind(Vdst)
+    L = length(uncoupleds_src_unique)
+    TStrided = StridedStructure{N}
+    data = Vector{Tuple{Matrix{T},Vector{TStrided},Vector{TStrided}}}(undef, L)
 
-    for (row, (f1, f2)) in enumerate(structure_src.fusiontreelist)
-        for ((f3, f4), coeff) in transform(f1, f2)
-            col = structure_dst.fusiontreeindices[(f3, f4)]
-            push!(rows, row)
-            push!(cols, col)
-            push!(vals, coeff)
+    # TODO: this can be multithreaded
+    for (i, uncoupled) in enumerate(uncoupleds_src_unique)
+        ids_src = findall(==(uncoupled), uncoupleds_src)
+        fusiontrees_outer_src = structure_src.fusiontreelist[ids_src]
+
+        uncoupled_dst = TupleTools.getindices(uncoupled, (p[1]..., p[2]...))
+        ids_dst = findall(==(uncoupled_dst), uncoupleds_dst)
+
+        fusiontrees_outer_dst = structure_dst.fusiontreelist[ids_dst]
+
+        matrix = zeros(sectorscalartype(I), length(ids_dst), length(ids_src))
+        for (row, (f₁, f₂)) in enumerate(fusiontrees_outer_src)
+            for ((f₃, f₄), coeff) in transform(f₁, f₂)
+                col = findfirst(==((f₃, f₄)), fusiontrees_outer_dst)::Int
+                matrix[row, col] = coeff
+            end
         end
+        @debug("Created recoupling block for uncoupled: $uncoupled",
+               sz = size(matrix), sparsity = count(!iszero, matrix) / length(matrix))
+
+        data[i] = (matrix,
+                   structure_dst.fusiontreestructure[ids_dst],
+                   structure_src.fusiontreestructure[ids_src])
     end
 
-    if FusionStyle(I) isa UniqueFusion
-        return AbelianTreeTransformer(rows, cols, vals, structure_dst, structure_src)
-    else
-        ldst = length(structure_dst.fusiontreelist)
-        lsrc = length(structure_src.fusiontreelist)
-        matrix = sparse(rows, cols, vals, ldst, lsrc)
-        return GenericTreeTransformer(matrix, structure_dst, structure_src)
+    # sort by (approximate) weight to make the buffers happy
+    # and use round-robin strategy for multi-threading
+    sort!(data; by=_transformer_weight, rev=true)
+
+    @debug("TreeTransformer for $Vsrc to $Vdst via $p",
+           nblocks = length(data),
+           sz_median = size(data[end ÷ 2][1], 1),
+           sz_max = size(data[1][1], 1))
+
+    return GenericTreeTransformer{T,N}(data)
+end
+
+# Cost model for transforming a set of subblocks with fixed uncoupled sectors:
+# L x L x size(subblock) where L is the number of subblocks
+# this is L input blocks each going to L output blocks of given size
+function _transformer_weight((matrix, structures_dst, structures_src))
+    return length(matrix) * prod(structures_dst[1][1])
```

Comment:
It might be that the main cost of performing the work for a single set of uncoupled charges is the `L` permutations, which involves data transfer of `L x length(subblock)` elements (by lack of a better cost model), and not the matrix multiplication (which indeed has cost `L x L x length(subblock)`).